File: day2/solve.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func readFileAsString(path string) string {
	b, err := os.ReadFile(path)
	if err != nil {
	}

	return string(b)
}

func calcLine(l string) int {
	split := strings.Split(l, ":")
	gameId, _ := strconv.Atoi(strings.Split(split[0], " ")[1])

	maxRed, maxBlue, maxGreen := 0, 0, 0

	sets := strings.Split(split[1], ";")

	for _, v := range sets {
		colors := strings.Split(v, ",")

		for _, v := range colors {
			countColor := strings.Split(strings.Trim(v, " "), " ")

			count, _ := strconv.Atoi(countColor[0])
			color := countColor[1]

			if color == "red" && count > maxRed {
				maxRed = count
			}

			if color == "blue" && count > maxBlue {
				maxBlue = count
			}

			if color == "green" && count > maxGreen {
				maxGreen = count
			}
		}
	}

	if maxRed > 12 || maxGreen > 13 || maxBlue > 14 {
		return 0
	}

	return gameId
}

func solve(lines []string) int {
	sum := 0

	for _, v := range lines {
		if v == "" {
			break
		}

		sum += calcLine(v)
	}

	return sum
}

func main() {
	input := readFileAsString("input")
	lines := strings.Split(input, "\n")

	fmt.Println(solve(lines))
}

File: day2/solve.hs
import Data.List.Split
import Data.List

type RGB = (Int, Int, Int)

main :: IO ()
main = do
  input <- filter (/= "") . splitOn "\n" <$> readFile "input"

  print(calcSet "3 blue, 4 red")

  print $ calcLine(head input)

  solve input

  where 
    solve :: [String] -> IO()
    solve lines = do 
      let x = foldl' step 0 lines
      print x
      
      where
        step :: Int -> String -> Int
        step acc line
          | r <= 12 && g <= 13 && b <= 14 = acc + id
          | otherwise =  acc
          where
            (id, (r,g,b)) = calcLine line

    calcLine :: String -> (Int, RGB)
    calcLine l = (read id :: Int, foldl' step (0, 0 ,0) gameSets)
      where
        [gameId, stringSets] = splitOn ": " l

        [_, id] = splitOn " " gameId

        gameSets = splitOn "; " stringSets

        step :: RGB -> String -> RGB 
        step (aR,aG,aB) gameSet = (
            max aR cR,
            max aG cG,
            max aB cB
          )
          where (cR,cG,cB) = calcSet gameSet

    calcSet :: String -> RGB
    calcSet set = foldl' step (0,0,0) valueColorList
      where 
        valueColorList = splitOn ", " set

        step :: RGB -> String -> RGB
        step (red, green, blue) curr
          | c == "red" = (read v :: Int, green, blue)
          | c == "green" = (red, read v :: Int, blue)
          | otherwise = (red, green, read v :: Int)

          where
            [v, c] = splitOn " " curr



